edx/edx-ora2
# Update filename encoding to handle more characters

**TL;DR -** Having a comma in a filename was causing download issues in Chrome. Update file name encoding to handle a broader range of special characters.

JIRA: [EDUCATOR-5592](https://openedx.atlassian.net/browse/EDUCATOR-5592)

**What changed?**

- Switch `encodeURI` to `encodeURIComponent` to encode additional characters, like the offending `,`
- Add additional encoding logic from this [MDN-suggested solution](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#encoding_for_content-disposition_and_link_headers)
- Add JS test for filename encoding

**Note:** I've tested this with a wide set of special characters, none of which I've tested break download/upload, but not all survive the encoding process. Specifically, I've noticed `|`, `?`, `~`, and `*` appear on the file link but not in the downloaded file. We may add this to any file info text on the page, something like "avoid using special characters in filenames, as they may not show up properly in downloads."

e.g. a file called `tëšt|,png*.png` will download as `tëšt_,png_.png`.

**Developer Checklist**


**Testing Instructions**

1. Configure testing environment to use S3 or an S3 stand-in (see [devstack/nsprenkle/localstack](https://github.com/edx/devstack/tree/nsprenkle/localstack))
2. In Chrome, upload files with different special characters and non-ASCII characters, for example `tëšt|^``#,;?@&=+$-_.!~*'().png`
3. Verify that upload/downloads work with (mostly) preserved filenames
4. Repeat on different browsers

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:


FYI: @edx/masters-devs-gta
